tests/reproducibility/test_graphsage.py

```diff
@@ -145,6 +145,67 @@ def gs_nai(
     return model
 
 
+def gs_link_pred_model(num_samples, generator, optimizer, bias, dropout, normalize):
+    layer_sizes = [50] * len(num_samples)
+    graphsage = GraphSAGE(
+        layer_sizes=layer_sizes,
+        generator=generator,
+        bias=bias,
+        dropout=dropout,
+        normalize=normalize,
+    )
+    # Build the model and expose input and output sockets of graphsage, for node pair inputs:
+    x_inp, x_out = graphsage.build()
+    pred = link_classification(
+        output_dim=1, output_act="relu", edge_embedding_method="ip"
+    )(x_out)
+
+    model = tf.keras.Model(inputs=x_inp, outputs=pred)
+
+    model.compile(
+        optimizer=optimizer, loss=tf.keras.losses.binary_crossentropy,
+    )
+
+    return model
+
+
+def gs_link_prediction(
+    g,
+    edge_ids,
+    edge_labels,
+    num_samples,
+    optimizer,
+    batch_size=4,
+    epochs=4,
+    bias=True,
+    dropout=0.0,
+    normalize="l2",
+    seed=0,
+    shuffle=True,
+):
+    set_seed(seed)
+    tf.random.set_seed(seed)
+    if shuffle:
+        random.seed(seed)
+
+    generator = GraphSAGELinkGenerator(g, batch_size, num_samples)
+    train_gen = generator.flow(edge_ids, edge_labels, shuffle=True)
+
+    model = gs_link_pred_model(
+        num_samples, generator, optimizer, bias, dropout, normalize
+    )
+
+    model.fit_generator(
+        train_gen,
+        epochs=epochs,
+        verbose=1,
+        use_multiprocessing=False,
```

Comment:
Based on the docs, I suspect `workers=4` is ignored if `use_multiprocessing=False`:

> workers: Integer. Used for generator or keras.utils.Sequence input only. Maximum number of processes to spin up **when using process-based threading**. ...
>
> use_multiprocessing: Boolean. Used for generator or keras.utils.Sequence input only. **If True, use process-based threading**. ...

(Emphasis mine)

Reply:
I was under the impression that `workers` refers to number of threads by default, but switches to number of processes if using multiprocessing. The docs aren't worded very well IMO but that's the behaviour I thought I saw.

> workers: Integer. Used for generator or keras.utils.Sequence input only. Maximum number of processes to spin up when using process-based threading. If unspecified, workers will default to 1. If 0, will execute the generator on the main thread.

Reply:
Ah, yeah, it seems you're right: https://github.com/tensorflow/tensorflow/blob/eb3dc469b0bf00f770960b1a265beb05af244083/tensorflow/python/keras/utils/data_utils.py#L732 . The `get_pool_class` function returns a `Pool` from either `multiprocessing` or [`multiprocessing.dummy`](https://docs.python.org/3.8/library/multiprocessing.html#module-multiprocessing.dummy) as appropriate.